wallet-new/integration/Functions.hs

```diff
@@ -0,0 +1,92 @@
+module Functions where
+
+import Universum
+
+import Control.Lens (ix)
+import Control.Monad.Random
+import Test.QuickCheck
+
+import Cardano.Wallet.API.V1.Types (Wallet (..), New)
+
+import Types
+import Error
+
+-- | The top function that we need to run in order
+-- to test the backend.
+runActionCheck
+    :: (WIntTestMode m)
+    => WalletState
+    -> ActionProbabilities
+    -> m WalletState
+runActionCheck walletState actionProb = do
+    action <- chooseAction actionProb
+    result <- runAction action
+
+    validateInvariants result walletState
+
+-- | Get action randomly, depending on the action distribution
+chooseAction
+    :: (WIntTestMode m)
+    => ActionProbabilities
+    -> m Action
+chooseAction aProb = do
```

Comment:
I'm not sure I understand this code entirely. It seems like it's starting with a probability distribution and it's generating a big list with a number of elements proportional to the probability. Example:

```haskell
[GetWallet, GetWallet, GetWallet, GetAccount, CreateTransaction, ...]
```

So that the more frequent the action the more the probability we pick it with `distributedActions ^? ix prob`.

Am I right or I'm still caffeine deprived?

I'm wondering if we are not overcomplicating this -- cannot we simply reify our probability map into a `Gen` and then use `frequency` from QuickCheck to achieve the same thing?

Reply:
You are correct.

I don't think we are overcomplicating this, the idea is to have an action distribution close to what a user has and to leave it running. After a while, the test will steer into some random state space that will be close to something a "real user" does.

Also, with those probabilities, we can include different "user profiles" (different action distributions) which can then test the application with the similar actions user from that profile would do. For example, we can have an "exchange user" and a "normal user" and test the application from those perspectives.
Not to mention that we can "translate" this into a Daedalus integration test (just an idea, but interesting nonetheless).

Reply:
My gripe is not with passing as input the `ActionProbabilities` (which can probably be renamed later into something which better reflect this idea of user profiles) but it should be totally possible to derive a `Gen` out of such data type (with a function of type `ActionProbabilities -> Gen Action`) which will compose way better with the way QuickCheck works.

Reply:
Oh, I see. I guess. I mean it really depends what we plan to use. I still think that using `IO` is the way to go since we do need the external interaction when communicating with the node, but we can use `Gen` if QuickCheck is going to be our "main weapon".
I don't see much difference myself, TBH, except the fact we are using something a bit heavier (and more powerful) like `IO`. 
I would go for this approach until we have a working version and then we can optimize.

Reply:
IO is almost never the way go for functions that don't need it. After all, we program in Haskell because we like things like purity, referential transparency, etc. Ideally a program is a thin side-effect-full (IO) layer on top of an otherwise pure program. QuickCheck's `Gen` monad is designed specifically for generating arbitrary data; it seems an obvious choice here -- not to mention that we may want to run these tests without any IO involved at all (if not right now, then in the future). Taking pure code and using in a stateful context is easy; starting with effectful code and attempting to use it in a pure context can often be very difficult. `IO` should never be the default starting point.

Reply:
Ok guys, will fix :smile: 

Regarding `IO` - I still feel it's very useful when you are prototyping something and don't want to "optimize prematurely". I don't know about you guys, but I rarely have everything "in my head". The task of programming for me is exploring the domain I'm in, and the speed at which I can do that is critical - not having to think about which Monad (domain) I'm in is a very useful thing. 
I rarely look at the code I put in the editor and go "clearly this is a MonadState with MonadV1 witch a pinch of MonadThatCantBeNeverFullyUnderstood". 
Then again, that's just me.

Reply:
Let me answer that question in earnest. Definitely, I agree both that premature optimization should be avoided and that when prototyping something we don't have all the details in our head and we don't yet fully know what the full context of this function will be. But here's the thing: types can _help_ with simplifying things. As soon as you declare your function to be in IO, basically all bets are off. _IO is as powerful as every other monad out there_ (including "`MonadState` with `MonadV1` and a pinch of `MonadThatCanNeverBeFullyUnderstood` :slightly_smiling_face:). When I prototype something, I start with the simplest type I can possibly imagine for this function, _and IO isn't it_. If I say `a -> a`, my function can only do one thing (must be the identity); if I say `Int -> Int`, I can do any operation on integers but nothing else; when I write `Gen Int`, I have a generator available. The point being that picking a simple type helps with making decisions. A type like `IO a` gives me no guidance at all because I can do _anything_. 

In fact, not only do I agree that premature optimization should be avoided, I agree that complicated types should be avoided whenever possible (this is an area where the Cardano codebase can be significantly improved). Understanding a function in `MonadThatCanNeverBeFullyUnderstood` indeed _is_ hard. It's one of the things that makes perusing the Cardano code base hard to newcomers (as well as making it hard to refactor it, or apply existing functions in different contexts, such as a pure unit testing context).

Keeping things simple is important; and _IO is not simple_.